packages/spacecat-shared-rum-api-client/src/functions/formVitals.js

```diff
@@ -0,0 +1,115 @@
+    if (!results[url]) {
+      results[url] = {
+        url,
+        formsubmit: {},
+        formview: {},
+        formengagement: {},
+      };
+    }
+
+    // Reference the current URL’s data object in the results
+    const urlData = results[url];
+    // Track if each condition has been processed for this event
+    const processedCheckpoints = new Set();
+
+    // Process each event within the bundle
+    for (const event of events) {
+      const { checkpoint, source } = event;
+
+      // Only process the checkpoint once per event
+      if (!processedCheckpoints.has(checkpoint)) {
+        // Check for 'viewblock' checkpoint indicating form views
+        if (checkpoint === 'viewblock' && FORM_SOURCE.includes(source)) {
+          accumulateByUserAgent(urlData.formview, userAgent, weight);
+          processedCheckpoints.add('viewblock'); // Mark as processed
+        }
+        // Check for 'formsubmit' checkpoint indicating form submissions
+        if (checkpoint === 'formsubmit') {
+          accumulateByUserAgent(urlData.formsubmit, userAgent, weight);
+          processedCheckpoints.add('formsubmit'); // Mark as processed
+        }
+        // Check for 'click' checkpoint with source indicating form engagement
+        if (checkpoint === 'click' && source && /\bform\b/.test(source.toLowerCase())) {
+          accumulateByUserAgent(urlData.formengagement, userAgent, weight);
+          processedCheckpoints.add('click'); // Mark as processed
+        }
+      }
+    }
+  }
+
+  return results;
+}
+
+function pageviewsByUrlAndUserAgent(bundles) {
+  return bundles.reduce((acc, cur) => {
+    const { userAgent } = cur;
+    if (!userAgent || userAgent.startsWith(BOT)) return acc;
+    // Initialize the URL object if it doesn't exist
+    acc[cur.url] = acc[cur.url] || {};
+    // Initialize the userAgent count if it doesn't exist
+    acc[cur.url][userAgent] = (acc[cur.url][userAgent] || 0) + cur.weight;
+    return acc;
+  }, {});
+}
+
+function handler(bundles) {
+  const pageviews = pageviewsByUrlAndUserAgent(bundles);
+  const formVitals = collectFormVitals(bundles);
+
+  return Object.values(formVitals)
+    .map((acc) => {
+      acc.pageview = pageviews[acc.url];
+      return acc;
+    })
+    .filter((item) => {
+      // Calculate the sum of values in formView, formSubmit, and formEngagement
+      const formViewSum = Object.values(item.formview).reduce((sum, value) => sum + value, 0);
+      const formSubmitSum = Object.values(item.formsubmit).reduce((sum, value) => sum + value, 0);
+      // eslint-disable-next-line max-len
+      const formEngagementSum = Object.values(item.formengagement).reduce((sum, value) => sum + value, 0);
+
+      // Check if any of the sums is greater than zero
+      return formViewSum > 0 || formSubmitSum > 0 || formEngagementSum > 0;
+    });
+}
+
+export default {
+  handler,
+  checkpoints: ['viewblock', 'formsubmit', 'click', 'error', 'top'],
```

Comment:
@vdua 

>The only issue that I see here is (which I have observed locally as well) that, for some customers like hdfc (or maybe adobe) having all bundles in memory for a month leads to memory issues.

If we take whole bundles (i.e, including all the events), there's high chance of hitting the nodejs memory limits (around 1.4GB by default), so we've introduced `checkpoints` filtering to reduce the memory footprint of processing. With `checkpoints` we've not hit the memory limits for 30 days for any customer, including adobe.com. 
Please note that the lambda functions are typically given 4GB memory, so we should be good as long as we filter the checkpoints. 
For adobe.com, when I tried to crunch data for 180 days, I didn't hit the memory limit but the timeout limit as the processing took more than 15 minutes

>which I have observed locally as well

May I know, what is the duration you're trying to crunch the data and what checkpoints you've used and which customer?

>the cwv query also loads almost all the bundles in memory. How do you handle running that query for a period of 1 month ?

`cwv` query filters by `checkpoints: ['cwv-lcp', 'cwv-cls', 'cwv-inp', 'cwv-ttfb']`, so not all the events will be in memory, so we didn't hit the memory issue for 30 days.

Crunching 180 days of RUM data for adobe.com is one of a kind situation, so I handled that locally and uploaded the results to dynamo. But if we see more of such scenarios, we can do progressive/incremental processing of RUM data and combine the aggregates as necessary. But I think, we don't need to go that extent atm.

Reply:
We are capturing for 1 month only for now but going forward we will have it for 1 week. I tried to run this locally for HDFC Bank last month and the program crashed. Since we are capturing `top` checkpoint, it will contain all the bundles. But I think we will be fine with 1 week

Reply:
@vdua 
>  tried to run this locally for a customer last month and the program crashed. Since we are capturing top checkpoint, it will contain all the bundles.

All the bundles will be there always, irrespective of checkpoints, to calculate the pageviews accurately. We'll only filter the bundle.events array for checkpoints. So, in your case, may be the default nodejs memory (1.4GB) would have reached and failed, but our audit workers have 4GB memory, so should be fine IMHO.